pyqtorch/embed.py

```diff
@@ -0,0 +1,295 @@
+from logging import getLogger
+from typing import Any, Optional, Tuple
+
+from numpy.typing import ArrayLike, DTypeLike
+
+logger = getLogger(__name__)
+
+
+ARRAYLIKE_FN_MAP = {
+    "torch": ("torch", "tensor"),
+    "jax": ("jax.numpy", "array"),
+    "numpy": ("numpy", "array"),
+}
+
+
+DEFAULT_JAX_MAPPING = {
+    "mul": ("jax.numpy", "multiply"),
+    "sub": ("jax.numpy", "subtract"),
+    "div": ("jax.numpy", "divide"),
+}
+DEFAULT_TORCH_MAPPING: dict = {}
+DEFAULT_NUMPY_MAPPING = {
+    "mul": ("numpy", "multiply"),
+    "sub": ("numpy", "subtract"),
+    "div": ("numpy", "divide"),
+}
+
+DEFAULT_INSTRUCTION_MAPPING = {
+    "torch": DEFAULT_TORCH_MAPPING,
+    "jax": DEFAULT_JAX_MAPPING,
+    "numpy": DEFAULT_NUMPY_MAPPING,
+}
+
+
+class ConcretizedCallable:
+    """Transform an abstract function name and arguments into
+        a callable in a linear algebra engine which can be evaluated
+        using user input.
+    Arguments:
+        call_name: The name of the function.
+        abstract_args: A list of arguments to the function,
+                       can be numeric types for constants or strings for parameters
+        instruction_mapping: A dict containing user-passed mappings from a function name
+                            to its implementation.
+        engine_name: The name of the framework to use.
+        device: Which device to use.
+
+    Example:
+    ```
+    import torch
+
+    from pyqtorch.embed import ConcretizedCallable
+
+
+    In [11]: call = ConcretizedCallable('sin', ['x'], engine_name='numpy')
+    In [12]: call({'x': 0.5})
+    Out[12]: 0.479425538604203
+
+    In [13]: call = ConcretizedCallable('sin', ['x'], engine_name='torch')
+    In [14]: call({'x': torch.rand(1)})
+    Out[14]: tensor([0.5531])
+
+    In [15]: call = ConcretizedCallable('sin', ['x'], engine_name='jax')
+    In [16]: call({'x': 0.5})
+    Out[16]: Array(0.47942555, dtype=float32, weak_type=True)
+    ```
+
+
+
+    """
+
+    def __init__(
+        self,
+        call_name: str,
+        abstract_args: list[str | float | int],
+        instruction_mapping: dict[str, Tuple[str, str]] = dict(),
+        engine_name: str = "torch",
+        device: str = "cpu",
+        dtype: Any = None,
+    ) -> None:
+        instruction_mapping = {
+            **instruction_mapping,
+            **DEFAULT_INSTRUCTION_MAPPING[engine_name],
+        }
+        self.call_name = call_name
+        self.abstract_args = abstract_args
+        self.engine_name = engine_name
+        self._device = device
+        self._dtype = dtype
+        self.engine_call = None
+        engine_call = None
+        engine = None
+        try:
+            engine_name, fn_name = ARRAYLIKE_FN_MAP[engine_name]
+            engine = import_module(engine_name)
+            self.arraylike_fn = getattr(engine, fn_name)
+        except (ModuleNotFoundError, ImportError) as e:
+            logger.error(f"Unable to import {engine_call} due to {e}.")
```

Comment:
Not sure this is `engine_call` ? It is `None` at this point.

Reply:
i made it more clear